### PR TITLE
Add station url requst  to a new thread.

### PR DESCRIPTION
**Pull request by sigmaSd**
clicking play button freeze the whole app for a few seconds , so I offloaded the url request into a new thread.
Works pretty well. 
Not sure if there is  a better way though.

### CODE REVIEW

File: rustio/src/client.rs [hunk 1]
```diff
@@ -93,10 +97,16 @@ impl Client {
         }
     }
 
-    pub fn get_playable_station_url(&self, station: &Station) -> String{
+    pub fn play_station(&self, station: &Station,mut atomic_playbin: Mutex<gstreamer::Element>) {
```
Comment (haecker-felix) [inline rustio/src/client.rs]: Rustio and Gradio are two crates, which are completely independently to each other. The rustio crate can be used by other applications too.

So it doesn't make sense here to return a gstreamer element. It forces the other application to use gstreamer too.

Rustio is just a API abstraction to radio-browser.info.

**Comment by sigmaSd** [comment]
I applied a new method but I doubt you will agree with it , so can I ask for some pointers on how to do the url request in an async way , or its too much trouble to add it now ?

File: rustio/src/client.rs [hunk 2]
```diff
@@ -35,14 +36,15 @@ pub enum ClientUpdate {
     Clear,
 }
 
+#[derive(Clone)]
 pub struct Client {
-    current_search_id: Rc<RefCell<u64>>,
+    current_search_id: Arc<Mutex<u64>>,
```
Comment (haecker-felix) [inline rustio/src/client.rs]: Why the change from Rc/RefCell to Arc/Mutex? I'm missing something?

File: src/audioplayer.rs [hunk 3]
```diff
@@ -63,13 +65,20 @@ impl AudioPlayer{
...
+        //request url and set it in a new thread
+        let atomic_playbin = Mutex::new(self.playbin.clone());
+        let atomic_client = self.client.clone();
+        let atomic_station = station.clone(); 
+        self.station=Some(station);
+        thread::spawn( move || {
+            let station_url = atomic_client.get_playable_station_url(&atomic_station);      
+            atomic_playbin.lock().unwrap().set_property("uri", &station_url);  
+            atomic_playbin.lock().unwrap().set_state(gstreamer::State::Playing);
+        });
```
Comment (haecker-felix) [inline src/audioplayer.rs]: Simple solution. I think this should be fine.

File: rustio/src/client.rs (hunk 2 above)
Comment (sigmaSd) [reply rustio/src/client.rs]: Needed to change it to be able to share the Client structure to the new thread, without those changes I get this
``
thread::spawn( move || {
   |         ^^^^^^^^^^^^^ `std::rc::Rc<std::cell::RefCell<u64>>` cannot be sent between threads safely
``
Thought you wouldn't agree cause each time I use clone multiple time I feel that maybe there's a better way plus didn't really want to change rustio crate at all but this is the solution I found for now.
Thanks for the feedback !

Comment (haecker-felix) [reply rustio/src/client.rs]: Oh, I see. 

I have no problem, if you edit the rustio crate. Both crates are still in development. Everything is fine, as long they are independently to each other.

Currently I'm on holiday, as soon as I'm home again, I'll merge your branch. Thanks for your work!

Comment (sigmaSd) [reply rustio/src/client.rs]: Thank you too I'm learning from your work plus Gradio is a very awesome app !

File: src/audioplayer.rs [hunk 4]
```diff
@@ -63,13 +65,20 @@ impl AudioPlayer{
...
-        self.station = Some(station);
-
         self.playbin.set_state(gstreamer::State::Null);
-        self.playbin.set_property("uri", &station_url);
+        self.station = Some(station);
+        
+        //request url and set it in a new thread
+        let playbin = Mutex::new(self.playbin.clone());
+        let client = self.client.clone();
+        let station = self.station.unwrap().clone();
```
Comment (haecker-felix) [inline src/audioplayer.rs]: I've renamed them, becuase as far I know this aren't real atomic types.[ More information here](https://doc.rust-lang.org/std/sync/atomic/index.html)

**Comment by haecker-felix** [comment]
Huh... Using AtomicUsize look more complicated than I thought.... I'll look if I can simplify it, otherwise it's better to use Arc/mutex.

**Comment by sigmaSd** [comment]
Its not actually complicated , the only thing extra job with this approach is to add clone function manually cause AtomicUsize doesn't have clone nor copy trait.
Arc<<Mutex has a performance overhead you should consider.
But Its true that its still feels like there could be a better approach to this problem.

**Comment by sigmaSd** [comment]
I'm curious can't we just use a simple u64 instead , why do you need Rc<<RefCell in the first place?
I tried it and it works normally .

**Comment by sigmaSd** [comment]
You're right , I played around with this , You have to keep track of the current_search with Arc<<Mutex , doesn't work  with u64 because it just gets copied around so I end up with search_id and current_search_id always the same and I cant send the original value to the timeout closure.

**Comment by sigmaSd** [comment]
Guess there is no need for heavy modifications!

**Comment by haecker-felix** [comment]
It seems everything is fine. I'm currently busy, so i cannot merge it right now. I want to test it before. It's not your fault. I'm sorry!

**Comment by sigmaSd** [comment]
Don't be mate , I'm just learning and trying out ideas you can accept or not its cool. 
Best of luck !

**Comment by haecker-felix** [comment]
Works flawless. Thanks for your work! I'm sorry, that I needed that long to merge it actually.

Feel free to continue to work on Gradio!

I found a pretty interesting piece of software, in the next days I'm going to to test this:
https://crates.io/crates/mdl

Maybe it can improve the code structure a bit :-)

**Comment by sigmaSd** [comment]
Thanks for your words :) It actually turned out to be better that it took so long since PR looks now way better than It looked at first.

> https://crates.io/crates/mdl

Looks Interesting , hope it turns out well !

**Comment by haecker-felix** [comment]
Yeah, looks very intresting, but all objects which should be "cached", needs to implement the serde Deserialize/Serialize trait, which is a bit hard for some objects... 

I also think it's better, if something needs longer, but it's well done then.

**Comment by haecker-felix** [comment]
@sigmaSd  first part of the rework is now pushed to master. The big advantages: 
- A signals/callback system
- The app_state/app_cache is now available everywhere, much easier to handle now (less Rc/RefCell code)
